### PR TITLE
fintune exec module code. and else/end opcode should not discard any …

### PR DESCRIPTION
**Pull request by carltraveler**

### CODE REVIEW

**Comment by carltraveler** [comment]
@sbinet  can you help review this?

**Comment by sbinet** [comment]
Just a quick message to say I haven't forgotten about this but I am teaching this week.
I'll circle back to this one before the end of the week.

**Comment by carltraveler** [comment]
ok, thanks

File: disasm/disasm.go [hunk 1]
```diff
@@ -129,13 +128,28 @@ func NewDisassembly(fn wasm.Function, module *wasm.Module) (*Disassembly, error)
...
 		logger.Printf("op: %s, unreachable: %v", opStr.Name, instr.Unreachable)
-		if !opStr.Polymorphic && !instr.Unreachable {
+		if !opStr.Polymorphic {
 			top := int(stackDepths.Top())
 			top -= len(opStr.Args)
 			stackDepths.SetTop(uint64(top))
-			if top < -1 {
-				return nil, ErrStackUnderflow
+			if top < 0 {
+				panic("should not happend underflow due to validation")
```
Comment (sbinet) [inline disasm/disasm.go]:
```suggestion
				panic("underflow during validation")
```

File: disasm/disasm.go [hunk 2]
```diff
@@ -230,17 +213,14 @@ func NewDisassembly(fn wasm.Function, module *wasm.Module) (*Disassembly, error)
 			// as well. To make sure that isInstrReachable
 			// returns the correct value, we don't push a new
 			// array to blockPolymorphicOps.
-			if !instr.Unreachable {
-				// Therefore, only push a new array if this instruction
-				// is reachable.
-				blockPolymorphicOps = append(blockPolymorphicOps, []int{})
-			}
+
+			// Therefore, only push a new array if this instruction
```
Comment (sbinet) [inline disasm/disasm.go]: I guess this comment block as well as the one above need some rewording in light of the control flow change that is being applied in this PR.

File: disasm/disasm.go [hunk 3]
```diff
@@ -318,44 +299,36 @@ func NewDisassembly(fn wasm.Function, module *wasm.Module) (*Disassembly, error)
...
-					if module.Types == nil {
-						return nil, errors.New("missing types section")
-					}
-					sig = &module.Types.Entries[index]
-					top--
-				} else {
-					sig = module.GetFunction(int(index)).Sig
+			var sig *wasm.FunctionSig
+			top := int(stackDepths.Top())
+			if op == ops.CallIndirect {
```
Comment (sbinet) [inline disasm/disasm.go]: I'd use a `switch` here:
```go
switch op {
case ops.CallIndirect: // ...
default: // ...
}
```

File: wasm/index.go [hunk 4]
```diff
@@ -98,6 +99,26 @@ func (m *Module) GetFunction(i int) *Function {
 	return &m.FunctionIndexSpace[i]
 }
 
+func (m *Module) GetFunctionSig(i uint32) *FunctionSig {
+	var funcindex uint32
+	if m.Import != nil {
+		for _, importEntry := range m.Import.Entries {
+			if importEntry.Type.Kind() == ExternalFunction {
```
Comment (sbinet) [inline wasm/index.go]: please inverse the boolean test and dedent the block.

File: wasm/index.go [hunk 5]
```diff
@@ -118,6 +139,24 @@ func (m *Module) GetGlobal(i int) *GlobalEntry {
 	return &m.GlobalIndexSpace[i]
 }
 
+func (m *Module) GetGlobalType(i uint32) *GlobalVar {
+	var globalindex uint32
+	if m.Import != nil {
+		for _, importEntry := range m.Import.Entries {
+			if importEntry.Type.Kind() == ExternalGlobal {
```
Comment (sbinet) [inline wasm/index.go]: please inverse the boolean condition and dedent the block.

File: wasm/index.go [hunk 6]
```diff
@@ -98,6 +99,26 @@ func (m *Module) GetFunction(i int) *Function {
 	return &m.FunctionIndexSpace[i]
 }
 
+func (m *Module) GetFunctionSig(i uint32) *FunctionSig {
```
Comment (sbinet) [inline wasm/index.go]: please add a test for this new exposed method.

File: wasm/index.go [hunk 7]
```diff
@@ -118,6 +139,24 @@ func (m *Module) GetGlobal(i int) *GlobalEntry {
 	return &m.GlobalIndexSpace[i]
 }
 
+func (m *Module) GetGlobalType(i uint32) *GlobalVar {
```
Comment (sbinet) [inline wasm/index.go]: please add a test for this new exposed method.

File: wasm/module_test.go [hunk 8]
```diff
@@ -173,3 +174,93 @@ func TestDuplicateExportError_NoStackOverflow(t *testing.T) {
...
+		t.Fatalf("%v", err)
+	}
+	defer f.Close()
+	m, err := wasm.ReadModule(f, nil)
+	if err != nil {
+		t.Fatalf("error reading module %v", err)
+	}
+
+	// check first sig
+	fsig, err := m.GetFunctionSig(0)
```
Comment (sbinet) [inline wasm/module_test.go]: these `m.GetFunctionSig(0,1,2,3)` could be rewritten as a table-driven test.
but ok, we can leave it for another PR :)
